# Step 1: Exercise for creating the “price_label” and “busiest_day” columns

In [6]:
# Import libraries
import pandas as pd
import os

# Define file paths
project_path = r'/Documents/15-01-2025 Instacart Basket Analysis'
prepared_data_path = os.path.join(project_path, '02 Data', 'Prepared Data')

# Load the ords_prods_merge dataframe
ords_prods_merge = pd.read_pickle(os.path.join(prepared_data_path, 'ords_prods_merge.pkl'))

# Confirm the data was loaded successfully
print(ords_prods_merge.head())


FileNotFoundError: [Errno 2] No such file or directory: '/Documents/15-01-2025 Instacart Basket Analysis/02 Data/Prepared Data/ords_prods_merge.pkl'

In [8]:
# Direct absolute path to the file
ords_prods_merge = pd.read_pickle('/Users/dela/Documents/15-01-2025 Instacart Basket Analysis/02 Data/Prepared Data/ords_prods_merge.pkl')

# Verify the data was loaded
print(ords_prods_merge.head())


   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order product_id  add_to_cart_order  reordered  \
0                     NaN        196                  1          0   
1                     NaN      14084                  2          0   
2                     NaN      12427                  3          0   
3                     NaN      26088                  4          0   
4                     NaN      26405                  5          0   

                              product_name  aisle_id  department_id  prices  
0                                     Soda  

# Create a subset of the dataframe

In [10]:
# Create a subset of the dataframe with the first 1,000,000 rows
df_subset = ords_prods_merge[:1000000]


# Apply the price_label Function

In [12]:
# Define the price_label function
def price_label(row):
    if row['prices'] <= 5:
        return 'Low-range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid-range product'
    elif row['prices'] > 15:
        return 'High-range product'
    else:
        return 'Not enough data'


In [14]:
# Apply the price_label function to the subset
df_subset['price_label'] = df_subset.apply(price_label, axis=1)

# Check the new column
print(df_subset['price_label'].value_counts())


price_label
Mid-range product     673189
Low-range product     314386
High-range product     12425
Name: count, dtype: int64


/var/folders/vr/vlcc2lxn4d7cl51gd3xv8gwh0000gn/T/ipykernel_4379/2198804588.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_subset['price_label'] = df_subset.apply(price_label, axis=1)


In [16]:
# Create a subset of the dataframe with the first 1,000,000 rows
df_subset = ords_prods_merge[:1000000].copy()

# Apply the price_label function to the subset
df_subset['price_label'] = df_subset.apply(price_label, axis=1)

# Check the new column
print(df_subset['price_label'].value_counts())


price_label
Mid-range product     673189
Low-range product     314386
High-range product     12425
Name: count, dtype: int64


# Create the busiest_day Column

In [18]:
# Create an empty list to store results
result = []

# Loop through the 'order_day_of_week' column and assign labels
for value in df_subset['order_day_of_week']:
    if value == 0:  # Saturday
        result.append("Busiest day")
    elif value == 4:  # Wednesday
        result.append("Least busy")
    else:
        result.append("Regularly busy")

# Add the list to the dataframe as a new column
df_subset['busiest_day'] = result

# Check the value counts of the new column
print(df_subset['busiest_day'].value_counts())


busiest_day
Regularly busy    692458
Busiest day       189674
Least busy        117868
Name: count, dtype: int64


# Compare Alternatives 

In [20]:
# Define the conditions and assign labels
df_subset.loc[df_subset['order_day_of_week'] == 0, 'busiest_day'] = "Busiest day"
df_subset.loc[df_subset['order_day_of_week'] == 4, 'busiest_day'] = "Least busy"
df_subset.loc[(df_subset['order_day_of_week'] != 0) & (df_subset['order_day_of_week'] != 4), 'busiest_day'] = "Regularly busy"

# Check the value counts of the new column
print(df_subset['busiest_day'].value_counts())


busiest_day
Regularly busy    692458
Busiest day       189674
Least busy        117868
Name: count, dtype: int64


# Remove the Subset and Use the Full Dataframe

In [22]:
# Reload the full dataframe
ords_prods_merge = pd.read_pickle('/Users/dela/Documents/15-01-2025 Instacart Basket Analysis/02 Data/Prepared Data/ords_prods_merge.pkl')

# Confirm the data loaded successfully
print(ords_prods_merge.head())


   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order product_id  add_to_cart_order  reordered  \
0                     NaN        196                  1          0   
1                     NaN      14084                  2          0   
2                     NaN      12427                  3          0   
3                     NaN      26088                  4          0   
4                     NaN      26405                  5          0   

                              product_name  aisle_id  department_id  prices  
0                                     Soda  

In [24]:
# Add the 'price_label' column to the full dataframe
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_label'] = 'Low-range product'
ords_prods_merge.loc[(ords_prods_merge['prices'] > 5) & (ords_prods_merge['prices'] <= 15), 'price_label'] = 'Mid-range product'
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_label'] = 'High-range product'

# Check the distribution of the new column
print(ords_prods_merge['price_label'].value_counts())


price_label
Mid-range product     21860852
Low-range product     10125759
High-range product      417678
Name: count, dtype: int64


In [26]:
# Add the 'busiest_day' column to the full dataframe
ords_prods_merge.loc[ords_prods_merge['order_day_of_week'] == 0, 'busiest_day'] = 'Busiest day'
ords_prods_merge.loc[ords_prods_merge['order_day_of_week'] == 4, 'busiest_day'] = 'Least busy'
ords_prods_merge.loc[(ords_prods_merge['order_day_of_week'] != 0) & 
                     (ords_prods_merge['order_day_of_week'] != 4), 'busiest_day'] = 'Regularly busy'

# Check the distribution of the new column
print(ords_prods_merge['busiest_day'].value_counts())


busiest_day
Regularly busy    22416495
Busiest day        6204040
Least busy         3783754
Name: count, dtype: int64


# Verify the Changes

In [28]:
print(ords_prods_merge.head())

print(ords_prods_merge['price_label'].value_counts())
print(ords_prods_merge['busiest_day'].value_counts())


   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order product_id  add_to_cart_order  reordered  \
0                     NaN        196                  1          0   
1                     NaN      14084                  2          0   
2                     NaN      12427                  3          0   
3                     NaN      26088                  4          0   
4                     NaN      26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda 

In [30]:
# Save the updated dataframe to a new pickle file
ords_prods_merge.to_pickle('/Users/dela/Documents/15-01-2025 Instacart Basket Analysis/02 Data/Prepared Data/ords_prods_merge_updated.pkl')


# Step 2: Update "Busiest day" to "Busiest days" and Identify Two Busiest and Two Slowest Days

In [32]:
# Find the frequency of orders by day
day_frequency = ords_prods_merge['order_day_of_week'].value_counts()
print(day_frequency)


order_day_of_week
0    6204040
1    5660135
6    4496403
2    4213760
5    4205721
3    3840476
4    3783754
Name: count, dtype: int64


In [34]:
# Define busiest and slowest days based on the frequency analysis
busiest_days = [0, 1]  # Top 2 busiest day indices
slowest_days = [4, 5]  # Bottom 2 slowest day indices

# Add the new "busiest_days" column
ords_prods_merge.loc[ords_prods_merge['order_day_of_week'].isin(busiest_days), 'busiest_days'] = 'Busiest days'
ords_prods_merge.loc[ords_prods_merge['order_day_of_week'].isin(slowest_days), 'busiest_days'] = 'Slowest days'
ords_prods_merge.loc[~ords_prods_merge['order_day_of_week'].isin(busiest_days + slowest_days), 'busiest_days'] = 'Regular days'


# Step 3: Check the Values for Accuracy and Note Observations

In [36]:
# Check value counts for the new column
print(ords_prods_merge['busiest_days'].value_counts())


busiest_days
Regular days    12550639
Busiest days    11864175
Slowest days     7989475
Name: count, dtype: int64


### Observations for Busiest Days

#### Value Distribution:
- **Regular days:** 12,550,639 rows
- **Busiest days:** 11,864,175 rows
- **Slowest days:** 7,989,475 rows

---

#### Accuracy Checks:
1. The **busiest days** label has been correctly applied to `order_day_of_week` values of `0` and `1`.
2. The **slowest days** label has been correctly applied to `order_day_of_week` values of `4` and `5`.
3. All remaining rows are labeled as **regular days**.

---

#### Trends Observed:
1. **Busiest Days:** Account for **37%** of the data.
2. **Slowest Days:** Represent **25%** of the data.
3. **Regular Days:** Comprise the remaining **38%**.

---

#### Insights:
1. **Marketing Opportunities:** Focus promotions on **regular days**.
2. **Operational Planning:** Prepare for high demand on the busiest days.
3. **Customer Behavior:** Midweek slowdowns might align with customer work schedules.


# step 4 : Label Busiest Periods of the Day

In [38]:
# Check the frequency of orders by hour
hour_frequency = ords_prods_merge['order_hour_of_day'].value_counts().sort_index()
print(hour_frequency)


order_hour_of_day
0      218766
1      115699
2       69374
3       51281
4       53241
5       87959
6      290492
7      891040
8     1718100
9     2454165
10    2761710
11    2736075
12    2618481
13    2660900
14    2689086
15    2662094
16    2535154
17    2087609
18    1636469
19    1258290
20     976145
21     795628
22     634216
23     402315
Name: count, dtype: int64


In [40]:
# Calculate thresholds
q25 = hour_frequency.quantile(0.25)  # Bottom 25%
q75 = hour_frequency.quantile(0.75)  # Top 25%

# Define the hours for each category
most_order_hours = hour_frequency[hour_frequency >= q75].index.tolist()
fewest_order_hours = hour_frequency[hour_frequency <= q25].index.tolist()


In [42]:
# Add "busiest_period_of_day" column
ords_prods_merge.loc[ords_prods_merge['order_hour_of_day'].isin(most_order_hours), 'busiest_period_of_day'] = 'Most orders'
ords_prods_merge.loc[ords_prods_merge['order_hour_of_day'].isin(fewest_order_hours), 'busiest_period_of_day'] = 'Fewest orders'
ords_prods_merge.loc[~ords_prods_merge['order_hour_of_day'].isin(most_order_hours + fewest_order_hours), 'busiest_period_of_day'] = 'Average orders'


# step5 : Print the frequency for this new column

In [46]:
# Print the frequency of the 'busiest_period_of_day' column
print(ords_prods_merge['busiest_period_of_day'].value_counts())


busiest_period_of_day
Most orders       16128346
Average orders    15679623
Fewest orders       596320
Name: count, dtype: int64


# Step 7: Export the DataFrame


In [48]:
# Define the export path
export_path = '/Users/dela/Documents/15-01-2025 Instacart Basket Analysis/02 Data/Prepared Data'

# Save the updated dataframe as a new version to avoid overwriting the existing file
ords_prods_merge.to_pickle(f'{export_path}/ords_prods_merge_updated_v2.pkl')

print("Dataframe saved successfully as ords_prods_merge_updated_v2.pkl in the Prepared Data folder!")


Dataframe saved successfully as ords_prods_merge_updated_v2.pkl in the Prepared Data folder!
